ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -443,6 +440,12 @@ public void replicate() {
         // still be able to handle duplicate response infos for the same request.
         responseInfos.addAll(responseInfosForTimedOutRequests);
         onResponses(responseInfos, correlationIdToRequestInfo, correlationIdToReplicaGroup);
+        if ((fastestReplicaFinishTime == null) && remoteReplicaGroups.stream().anyMatch(RemoteReplicaGroup::isDone)) {
+          fastestReplicaFinishTime = time.milliseconds();
+        }
+        remoteReplicaGroups.stream().filter(RemoteReplicaGroup::isDone).forEach(group -> {
+          groupFinishTime.putIfAbsent(group.getId(), time.milliseconds());
+        });
```

Comment:
We wanna look at the median and max idle time of a each batch, because replication runs in batches currently. Plotting individual idle times, I imagine, doesn't give us an accurate picture. n-1 replicas could finish together and so their idle-times would be small. While 1 replica finished first and the p90 would be close to it, which isn't quite right. 

Emit ReplicaGroupMedianIdleTimeSecs and ReplicaGroupMaxIdleTimeSecs - just two metrics per batch. That tells the number of seconds for which 50% of groups were idle in a batch.

Reply:
On second thought, to make it easier, you could sum up the idle time for a batch of ReplicaGroups.

Reply:
Think we can keep it simple, Sum(idle_time_of_individual_group)  divided by
((cycle_finish_time-cycle_start_time) * number_of_groups). And think it should be fine to have these emitted in milliseconds.

Reply:
I agree with keeping it simple and I think `Sum(idle_time_of_individual_group)` is simpler than `Sum(idle_time_of_individual_group) divided by ((cycle_finish_time-cycle_start_time) * number_of_groups)`.  What does the latter mean ?

Reply:
Among the total available time, how much of time was wasted being idle. 
Total available time = ((cycle_finish_time-cycle_start_time) * number_of_groups).

Reply:
`((cycle_finish_time-cycle_start_time) * number_of_groups)` is not the total time available because its not strictly a sequential execution i.e the loop is not waiting for one group to finish before moving to the next. The loop is sending out messages for one group after another without waiting for responses. It's parallel, sorta.
 
The actual total time is just `cycle_finish_time-cycle_start_time`, assuming these are times when the loop starts and ends. I see that you are trying to go for a percentage but the problem with percentages is that it does not tell us absolute value. For eg. if the outcome is that we wasted 50% of the time idling - which on the face of it sounds like a lot but isn't very much if the loop actually just took 100ms, and that's hidden.

Let's keep it simple for now and go for a sum. I am trying to get some clear visibility. If the total idle time is small, like a few millisecs, then its not worth optimizing replication. If the total idle time is large, in the order of seconds or minutes, then we can proceed to the next phase and optimize replication without chopping up the metric further or we could refine it if needed.

Reply:
Agreed on your point , I have added cumulative and percentage metrics both.
Even if cumulative metrics comes out to be 10 seconds, but if we have 1000 groups it will have less impact than if we have 1 second idle time for 1 group. But if cumulative is too less, we should also consider behaviour then. So I have added both metric so we have better data points.

Reply:
@snalli Get your point, but for the very same reason thought percentage could be helpful. Thought we are already having the metrics for cycle duration. As long as we have the data points for the aspects we are talking about and could take right decisions, it should be sufficient. Thanks!

Reply:
@skandasa23 Thanks, I understand now why its multiplied by number of groups.